### PR TITLE
add the support to specify device index for device_guard

### PR DESCRIPTION
**Pull request by sandyhouse**
PR types: New features
PR changes: Others
Describe: add the support for device index for device_guard, such as "gpu:1" to support pipeline.
Now, device index is only used for pipeline parallelism. For non-pipeline, the device index is ignored.

### CODE REVIEW

File: paddle/fluid/framework/op_proto_maker.cc [hunk 1]
```diff
@@ -89,6 +89,9 @@ void OpProtoAndCheckerMaker::operator()(proto::OpProto* proto,
       .SetDefault({});
   AddAttr<std::string>(OpDeviceAttrName(), "Device type of this operator.")
       .SetDefault("");
+  AddAttr<std::string>(OpDeviceIndexAttrName(),
+                       "Device index of this operator.")
+      .SetDefault("");
```
Comment (zhangting2020) [inline paddle/fluid/framework/op_proto_maker.cc]: 这里index的设置统一用op_device是不是python端能去掉很多修改？在c++端解析字符串如何？

Comment (sandyhouse) [reply paddle/fluid/framework/op_proto_maker.cc]: Done.

File: paddle/fluid/framework/operator.cc [hunk 2]
```diff
@@ -1050,7 +1050,15 @@ void OperatorWithKernel::ChooseKernel(const RuntimeContext& ctx,
...
       expected_kernel_key.place_ = platform::CPUPlace();
-    } else if (Attr<std::string>("op_device") == "gpu") {
+    } else if (Attr<std::string>("op_device").find("gpu") !=
+               std::string::npos) {
+      auto device = Attr<std::string>("op_device");
+      size_t pos = device.find(':');
+      if (pos != std::string::npos) {
+        device = device.substr(0, pos);
+        LOG_FIRST_N(WARNING, 1)
+            << "The device index for device_guard will be ignored.";
```
Comment (zhangting2020) [inline paddle/fluid/framework/operator.cc]: 如果支持了设置设备id，这里的warning是不是也会打印出来呢？明明是支持的，但是又提示设备索引会被忽略好像会比较困惑。

Comment (sandyhouse) [reply paddle/fluid/framework/operator.cc]: Done. Modify the warning message.

**Review by zhangting2020 — APPROVED** [review]
LGTM

**Review by Xreki — APPROVED** [review]
LGTM for the modification of framework.py

**Pull request merged**